src/d.ts

```diff
@@ -36,6 +36,28 @@ export function isHNode(child: DNode): child is HNode {
 	return Boolean(child && (typeof child !== 'string') && child.type === HNODE);
 }
 
+/**
+ * Generic decorate function for DNodes, the modified in place based on the provided predicate
+ * and modifier functions.
+ *
+ * If no predicate is supplied then the modifier will be executed on all nodes.
+ */
+export function decorate(dNodes: DNode | DNode[], modifier: (dNode: DNode) => void, predicate?: (dNode: DNode) => void): DNode | DNode[] {
+	let nodes = Array.isArray(dNodes) ? [ ...dNodes ] : [ dNodes ];
+	while (nodes.length) {
+		const node = nodes.pop();
+		if (node) {
+			if (!predicate || predicate(node)) {
```

Comment:
It just really wasn't obvious to me at all what this is doing or why. I mean, I understand the words decorate, predicate, and modifier, but I guess I'm lacking context for how this is used. Also it wasn't clear to me at a first glance, if something is either not a predicate or a predicate call with node as a parameter, why these are the conditions to call modifier on the node. And for the statement below, basically if something is a WNode or HNode and it has children, then a list of nodes is an array that is spread from a list of nodes and children? So basically you're flattening the loop of nodes to iterate over? And you're modifying the list of nodes within the while loop? I find it confusing to change the length of a while loop while iterating over the while loop, but that may just be me (there's no side effects as you're not adding to the list of nodes before the one you're currently iterating over obviously). :) Anyway, I'm probably just being too difficult on this particular block of code, but those were my thoughts when reviewing it.